SONARJAVA-4495 Update S2168 in line with LayC. (#2817)
SonarSource/rspec: rules/S2168/java/rule.adoc

== Why is this an issue?

> Double-checked locking is the practice of checking a lazy-initialized object's state both before
> and after a `synchronized` block is entered to determine whether to initialize the object.
> In early JVM versions, synchronizing entire methods was not performant, which sometimes
> caused this practice to be used in its place.

Reviewer: ```suggestion
Double-checked locking is the practice of checking a lazy-initialized object's state both before and after a `synchronized` block is entered to determine whether to initialize the object.
In early JVM versions, synchronizing entire methods was not performant, which sometimes caused this practice to be used in its place.
```

> Apart from `float` and `int` types, this practice does not work reliably in a platform-independent
> manner without additional synchronization of mutable instances.
> Using double-checked locking for the lazy initialization of any other type of primitive or mutable
> object risks a second thread using an uninitialized or partially initialized member while the first
> thread is still creating it.
> The results can be unexpected, potentially even causing the application to crash.

Reviewer: ```suggestion
Apart from `float` and `int` types, this practice does not work reliably in a platform-independent manner without additional synchronization of mutable instances.
Using double-checked locking for the lazy initialization of any other type of primitive or mutable object risks a second thread using an uninitialized or partially initialized member while the first thread is still creating it.
The results can be unexpected, potentially even causing the application to crash.
```

== How to fix it

> More recent JVM versions have significantly improved the performance of `synchronized` methods, so
> prefer using it over double-checked locking.
> Synchronized methods are now the preferred solution and are significantly more robust.

Reviewer: ```suggestion
Given the significant performance and robustness improvements of `synchronized` methods in more recent JVM versions, using them over double-checked locking is recommended.
```

> If marking the entire method as `synchronized` is not an option, consider using an inner
> `static class` to hold the reference instead.

Reviewer: ```suggestion
If marking the entire method as `synchronized` is not an option, consider using an inner `static class` to hold the reference instead.
```
